janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/util/RestClientAuthenticator.java

```diff
@@ -0,0 +1,28 @@
+package org.janusgraph.diskstorage.es.rest.util;
+
+import java.io.IOException;
+
+import org.elasticsearch.client.RestClientBuilder.HttpClientConfigCallback;
+import org.elasticsearch.client.RestClientBuilder.RequestConfigCallback;
+
+/**
+ * <p>
+ * A class instance implementing this interface may customize some aspects of
+ * JanusGraph ES REST client to implement a custom authentication method.
+ * </p>
+ * <p>
+ * A class implementing instance must provide a constructor that accepts a String[]. This
+ * array will contain the values provided as constructor parameters in JanusGraph graph index
+ * backend configuration. If the list of parameters is empty, a zero-size array is passed to the
+ * constructor. The constructor must not perform any blocking operations.
+ * </p>
+ */
+public interface RestClientAuthenticator extends HttpClientConfigCallback, RequestConfigCallback {
+
+    /**
+     * Do all your initialization (reading configuration files etc) here.
+     * @throws IOException
+     */
+    public void init() throws IOException;
```

Comment:
It might be clearer to implement the `String[]` args in init instead of the constructor.

Reply:
I would prefer to keep it this way unless there are strong arguments against it. This allows the constructor to fail early if some of the values are invalid or inconsistent. And then init() may fail if the second step of the initialization fails.

In fact, I have found that my AWS authenticator may violate my own rule because some of these SDK methods might do I/O under the hood. So I have updated the sample code as well :)

Reply:
Understood. Keeping it in the constructor makes sense.